File: src/parser/nim/actions/enable.nim
import os, asyncdispatch, json, sequtils, algorithm, strformat, strutils, re

import ../utils/[paths]

proc nlcli_enable*(s: string = "{}") {.async.} =
    let registrypath = paths["registrypath"]

    let jdata = parseJSON(s)
    let all = jdata{"all"}.getBool()
    var packages = toSeq(jdata{"_"}).mapIt(it.getStr())
    let action = packages[0].string
    packages.delete(0, 0)
    let state = if action == "enable": "false" else: "true"

    # Get all packages when '--all' is provided.
    if all:
        packages.setLen(0)
        packages = (
            const dirtypes = {pcDir, pcLinkToDir}
            for kind, path in walkDir(registrypath):
                if kind notin dirtypes: continue
                let parts = splitPath(path)
                packages.add(parts.tail)
            packages.sort()
            packages
        )

    # Loop over packages and remove each if its exists.
    for pkg in packages:
        let filepath = fmt"{registrypath}/{pkg}/.{pkg}.config.acdef"
        let resolved_path = absolutePath(filepath)
        # Workout symlinks: expandSymlink(), symlinkExists()

        if not fileExists(resolved_path): continue
        var contents = readFile(resolved_path).strip()

        contents = contents.replace(re("^@disable[^\n]*", {reMultiLine}), "").strip()
        contents &= fmt"\n@disable = {state}\n"
        contents = contents.replace(re("^\n/", {reMultiLine}), "") # Remove newlines.
        contents = contents.replace(re("\n", {reMultiLine}), "\n\n") # Add newline after header.

        writeFile(filepath, contents)

File: src/parser/js2/actions/enable.js
"use strict";

const flatry = require("flatry");
const fe = require("file-exists");
const toolbox = require("../utils/toolbox.js");
const { paths, read, write, realpath, readdir } = toolbox;

module.exports = async (args) => {
	let { registrypath } = paths;
	let { all } = args;
	let packages = args._;
	let action = packages[0];
	packages.shift();
	let state = action === "enable" ? "false" : "true";

	// Get all packages when '--all' is provided.
	if (all) packages = await readdir(registrypath);

	// Loop over packages and remove each if its exists.
	for (let i = 0, l = packages.length; i < l; i++) {
		let pkg = packages[i];

		let filepath = `${registrypath}/${pkg}/.${pkg}.config.acdef`;
		let [err, res] = await flatry(realpath(filepath));
		if (err) continue;
		let resolved_path = res;

		[err, res] = await flatry(fe(resolved_path));
		if (err || !res) continue;
		[err, res] = await flatry(read(resolved_path));
		if (err) continue;

		let contents = res.trim();
		contents = contents.replace(/^@disable[^\n]*/gm, "").trim();
		contents += `\n@disable = ${state}\n`;
		contents = contents.replace(/^\n/gm, ""); // Remove newlines.
		contents = contents.replace(/\n/, "\n\n"); // Add newline after header.

		[err] = await flatry(write(filepath, contents));
		if (err) continue;
	}
};
